### ISSUE TITLE
[bug] Infinite recursion when checking for an option in the package_id method

### ISSUE THREAD

**Issue by rddesmond** [conan-io/conan#7299]
Thank you for all the great work in conan!  I was working on a base class that would invoke `shared_library_package_id` conditionally if "shared" was in the options.  (A try-catch would work for this case, but I was working through some ideas.)  I found that the `in` lookup was causing infinite stack recursion.

```python
def package_id(self):
    if "not_an_option" in self.info.options:
        self.info.shared_library_package_id()
```

### Environment Details (include every applicable attribute)
  * Operating System+version: Ubuntu 16.04.5
  * Conan version: Conan version 1.27.0
  * Python version: Python 3.5.2

Also tested in the docker:
  * Python version: Python 3.7.5
  * Conan versions: 1.0.0, 1.10.0, 1.27.0
  * Docker container:  `conanio/gcc5`

### Steps to reproduce (Include if Applicable)

1. Create a new conanfile:
   ```bash
   conan new test/1.0@user/testing
   ```
1. Append a `package_id` method which looks for an option that doesn't exist:
   ```bash
   printf "    def package_id(self):\n        'not_an_option' in self.info.options" >> conanfile.py
   ```
1. Attempt to create a package:
   ```bash
   conan create . user/testing
   ```

**Expected:** Builds 
**Actual:** 100% CPU and eventually 100% RAM

```
top - 22:12:09 up  6:26,  0 users,  load average: 0.89, 0.55, 0.41
Tasks:   4 total,   2 running,   2 sleeping,   0 stopped,   0 zombie
%Cpu(s): 13.2 us,  0.8 sy,  0.0 ni, 86.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem : 16281688 total,  1775688 free, 10938668 used,  3567332 buff/cache
KiB Swap: 33248252 total, 30537368 free,  2710884 used.  4498560 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND     
  630 conan     20   0 1727356 1.554g   9544 R  99.7 10.0   0:11.36 conan       
    1 conan     20   0   18272   3408   2896 S   0.0  0.0   0:00.31 bash        
  313 conan     20   0   18272   3372   2864 S   0.0  0.0   0:00.28 bash        
  629 conan     20   0   36640   3156   2732 R   0.0  0.0   0:00.00 top         
```

### Logs (Executed commands with output) (Include/Attach if Applicable)

When running in pdb, I added a breakpoint at `conans/model/options.py:216` (line number for version 1.27.0), which led to an infinite loop there:

```
(Pdb) n
--Return--
> /home/rdesmond/.virtualenvs/conan-1.26.0/lib/python3.5/site-packages/conans/model/options.py(216)__getitem__()-><conans.model...x7f5c503ff5c0>
-> return self._reqs_options.setdefault(item, PackageOptionValues())
(Pdb) n
--Call--
> /home/rdesmond/.virtualenvs/conan-1.26.0/lib/python3.5/site-packages/conans/model/options.py(215)__getitem__()
-> def __getitem__(self, item):
(Pdb) n
> /home/rdesmond/.virtualenvs/conan-1.26.0/lib/python3.5/site-packages/conans/model/options.py(216)__getitem__()
-> return self._reqs_options.setdefault(item, PackageOptionValues())
(Pdb) n
--Return--
> /home/rdesmond/.virtualenvs/conan-1.26.0/lib/python3.5/site-packages/conans/model/options.py(216)__getitem__()-><conans.model...x7f5c503ffda0>
-> return self._reqs_options.setdefault(item, PackageOptionValues())
```

**Comment by memsharded** [comment]
Thanks for the detailed report! I have provided the necessary ``__contains__`` method so this check is possible: https://github.com/conan-io/conan/pull/7303

In any case, I would also suggest to reconsider the check. It sounds unusual to do the logic based on the existence of an option and not its value. Which option are you checking? Is it ``shared``? Maybe something that could be built-in in the ``shared_library_package_id()``?

**Comment by rddesmond** [comment]
Whoa, that was really fast, thank you!  I mirrored it after the check in `shared_library_package_id` itself (why doesn't it's check ` if "shared" not in dep_options` break too?), but with the goal of a base class having similar logic.  Specifically, this was a base class for boost, which has some header-only components and some with a compiled library.  

Boost has no requirements, so this probably shouldn't be there anyhow, but the logic was something along the lines of "if it's header-only the requirements never matter; if it's has a shared option they should follow the `shared_library_package_id`'s logic."

Either way, I was confused when my computer nearly fell over from standard python, so the fix is quite welcome!

```python
class ConanInfo(object):
    ...
    def shared_library_package_id(self):
        if self.full_options.shared:
            for dep_name in self.requires.pkg_names:
                dep_options = self.full_options[dep_name]
                if "shared" not in dep_options or not self.full_options[dep_name].shared:
                    self.requires[dep_name].package_revision_mode()
    ...
```

**Comment by memsharded** [comment]
Good point. This is because internally, it is checking against ``self.info.full_options``, and the attribute check there doesn't raise if the attribute (in this case the ``shared`` option) is not defined. Same happens if you directly test the ``self.info.options.whatever``, it will not fail if not existing. I have added those cases to the tests in the PR too.

**Comment by memsharded** [comment]
Implemented the ``in`` operator in #7303, together with a bug that didn't raise for wrong options. Will be in next 1.28, thanks for reporting!
